[Benjamin Lee] iP

BenjaminLHR opened the pull request:
## Duke
> "After using Duke, I was able to water my plants, plan a holiday, run a marathon and complete a PhD!" – Satisfied Customer

Duke frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/BenjaminLHR/ip).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you 😉

And it is **FREE**!

Features
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

This project was implemented in Java:
```java
public int getSize() {
    return listOfTasks.size();
}
```

Commit:
Add Gradle support

BenjaminLHR committed:
Create skeleton for Ui

BenjaminLHR committed:
Create skeleton for Command class

BenjaminLHR committed:
Create skeleton for Duke

BenjaminLHR committed:
Rename readCommand() to readInput()

BenjaminLHR committed:
Rename variables to be consistent with renaming readCommand() to readInput()

BenjaminLHR committed:
Add skeleton for TaskList

BenjaminLHR committed:
Implement ability to list tasks and modify default execute behaviour

BenjaminLHR committed:
Add argument to execute method call

BenjaminLHR committed:
Use a Task class instead of String to represent each task

BenjaminLHR committed:
Organize classes into packages (commands, core, gui, tasks)

BenjaminLHR committed:
Change Command to an abstract class and add its subclasses

BenjaminLHR committed:
Add Parser

BenjaminLHR committed:
Add functionality to mark task as completed

BenjaminLHR committed:
Reorder imports

BenjaminLHR committed:
Remove unnecessary import

BenjaminLHR committed:
Change Ui's methods to static

BenjaminLHR committed:
Create subclasses Todo, Deadline, Event that inherit from Task.

BenjaminLHR committed:
Create subclasses DeadlineCommand, EventCommand, TodoCommand that inherit from Command.

BenjaminLHR committed:
Change addTask display message

BenjaminLHR committed:
Modified switch cases to handle todo, deadline and event command types.

BenjaminLHR committed:
Fix minor typo in Ui.greet()

BenjaminLHR committed:
Set up text-ui-test that, with expected output based on Level-4

BenjaminLHR committed:
Add error handling

BenjaminLHR committed:
Add new method formatAndPrint

BenjaminLHR committed:
Use formatAndPrint

BenjaminLHR committed:
Add HelpfulFunctions

BenjaminLHR committed:
Add delete functionality

BenjaminLHR committed:
Add delete method to taskList

BenjaminLHR committed:
Revert "Add delete method to taskList"

This reverts commit 7c505780554b9590932e33d3bca63fc7b4245b0c.

BenjaminLHR committed:
Add delete method to taskList

BenjaminLHR committed:
Restructure exception checking and add more exception cases

BenjaminLHR committed:
Instantiate an empty duke.txt

BenjaminLHR committed:
Create new class Storage with addTasksToFile method

BenjaminLHR committed:
Add storage parameter to execute and make use of it when appropriate

BenjaminLHR committed:
Add storage argument to command.execute() method

BenjaminLHR committed:
Add saveContents method

BenjaminLHR committed:
Add toStorageFormat method

BenjaminLHR committed:
Add "data/duke.txt" to gitignore

BenjaminLHR committed:
Fix indentation for switch statements

BenjaminLHR committed:
Import commands individually instead of using wildcard

BenjaminLHR committed:
Add addTaskQuietly method

BenjaminLHR committed:
Change format of storage data

BenjaminLHR committed:
Add functionality to load tasks from "data/Duke.txt"

BenjaminLHR committed:
Add functionality that creates the /data directory if it does not exist

BenjaminLHR committed:
Add additional statement to be printed if "data/Duke.txt" cannot be found.

BenjaminLHR committed:
Rename attribute "date" to "time"

BenjaminLHR committed:
Add timeToString() method which allows Duke to understand dates and output them in a meaningful format

BenjaminLHR committed:
Add "data/duke.txt" to gitignore

BenjaminLHR committed:
Merge branch 'branch-Level-7'

BenjaminLHR committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	.gitignore

BenjaminLHR committed:
Modify Duke's constructor to require filePath

BenjaminLHR committed:
Add load() method and remove loadStorageToTaskList() method

BenjaminLHR committed:
Add a new constructor that accepts an ArrayList<Task> argument and makes use of it to instantiate listOfTasks

BenjaminLHR committed:
Modify instantiation of taskList to make use of storage.load()

BenjaminLHR committed:
Add showStorageFileNotFoundError() method

BenjaminLHR committed:
Add exception checking for load()

BenjaminLHR committed:
Merge branch 'branch-Level-7'

BenjaminLHR committed:
Fix bug in toStorageFormat() by changing "date" to "time"

BenjaminLHR committed:
Rename package commands to command

BenjaminLHR committed:
Rename package "tasks" to "task"

BenjaminLHR committed:
Put all classes under a "Duke" package

BenjaminLHR committed:
Initialize TaskListTest with a test method testGetSize()

BenjaminLHR committed:
Fix formatting

BenjaminLHR committed:
Initialize DeadlineTest with a test methods testStringConversion() and testStorageFormatConversion()

BenjaminLHR committed:
Initialize TodoTest with a test methods testStringConversion() and testStorageFormatConversion()

BenjaminLHR committed:
Change helper methods to private

BenjaminLHR committed:
Change helper methods that are only used within Storage to private

BenjaminLHR committed:
Add MANIFEST.MF to gitignore

BenjaminLHR committed:
Add new method -- findAndListTasks()

BenjaminLHR committed:
Initialize class FindCommand

BenjaminLHR committed:
Add basic functionality for handling "find" commands in Parser.parse()

BenjaminLHR committed:
Merge branch 'branch-Level-9'

BenjaminLHR committed:
Format code to align with the module's coding standard.

BenjaminLHR committed:
Merge branch 'branch-A-CodingStandard'

BenjaminLHR committed:
Add Javadoc comments

BenjaminLHR committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/command/ByeCommand.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/core/Storage.java
#	src/main/java/duke/core/TaskList.java
#	src/main/java/duke/gui/Ui.java

ErnestCuong commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,44 @@
+     * @return false
+     */
+    @Override
+    public boolean shouldExit() {
```
Perhaps you can name it as "isExit".

ErnestCuong reviewed (commented):
Looks good. There seems to be no naming violation.

myathk commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,43 @@
+ * Encapsulates a command to delete a task based on its index.
+ */
+public class DeleteCommand extends Command {
+    private int indexToDelete;
```
I like how the name of this private field is descriptive rather than having a generic name of just "index".

myathk commented on src/main/java/duke/core/Parser.java:
```diff
@@ -0,0 +1,159 @@
+
+    }
+
+    private static boolean checkValiditiy(String input, TaskList taskList) throws DukeException {
```
Alternatively, would it be better to have this method named as "isValid" instead as it is returning a boolean?

myathk commented on src/main/java/duke/core/TaskList.java:
```diff
@@ -0,0 +1,134 @@
+    public void delete(int index) {
+        Task taskToRemove = listOfTasks.get(index - 1);
+        listOfTasks.remove(index - 1);
+        String outputLine1 = "Noted. I've removed this task: \n";
+        String outputLine2 = taskToRemove.toString() + "\n";
+        String outputLine3 = String.format("Now you have %s tasks in the list.", listOfTasks.size());
+        String output = outputLine1 + outputLine2 + outputLine3;
```
Should these String variables have more descriptive names to differentiate between themselves?

myathk commented on src/main/java/duke/core/TaskList.java:
```diff
@@ -0,0 +1,134 @@
+        Ui.formatAndPrint(sb.substring(0, sb.length() - 1));
+    }
+
+    public void findAndListTasks(String keyword) {
```
I really like how this method is named, good job!

myathk reviewed (commented):
In general, the code follows the coding standard very closely and I really like your naming of variable names and methods. They are very descriptive and easy to make sense of.

BenjaminLHR committed:
Merge branch 'add-gradle-support' of https://github.com/BenjaminLHR/ip into A-Gradle

BenjaminLHR committed:
Merge tag 'A-Gradle'

BenjaminLHR committed:
Configure build.gradle to support JavaFX

BenjaminLHR committed:
Initialize a limited but functional GUI

BenjaminLHR committed:
Merge branch 'JavaFX'

BenjaminLHR committed:
Change return type of Command.execute() from void to String. Instead of printing the output directly, Command.execute() will instead return the output as a String.

BenjaminLHR committed:
Add a statement to print the output from command.execute() in run()

BenjaminLHR committed:
Add method getResponseAndExitStatus

BenjaminLHR committed:
Change emoticon representation for DukeException messages

BenjaminLHR committed:
Create GUi for Duke

BenjaminLHR committed:
Merge branch 'JavaFX'

BenjaminLHR committed:
Change mainClassName to "duke.gui.Launcher" so that the JAR file created
by gradle is that of Launcher.

BenjaminLHR committed:
Add assertions that check that the indexes of tasks to be deleted or
marked as complete are at least 1.

Should the value of the indexes be less than 1 at the point the
assertions are made be less than 1, the program would crash subsequently
because we would be trying to access a negative index of an ArrayList.
As a first line of defense, we have tried to use exceptions to warn
users should they enter inappropriate index values. The assertions are
put in place in case we (as programmers) fail to use exception handling
in certain cases.

Let's add assertions that terminate the program (and display messages
explaining the terminations) should the indexes be less than 1.

Using assertions is a good choice because their impact on performance
is low and worth the additional safety provided.

BenjaminLHR committed:
Define formatTime() method in Task instead of Deadline and Event to
follow the DRY principle.

BenjaminLHR committed:
Modify FindCommand's execute() method to return the output as a string
instead of displaying the output directly.

BenjaminLHR committed:
Enhance readability by using named constants for regexes and command
types.

BenjaminLHR committed:
Extract out checkNumOfArguments() and checkIndexArgument() from the
variations of checkValidityCommand() methods.

Certain behaviors are shared among several variations of
checkValidityCommand() methods, which violates the DRY principle.

Let's extract out the common behaviors into separate methods. For
example, all the checkValidityCommand() variations check for an
appropriate number of arguments in the input. Therefore, we abstracted
out a checkNumOfArguments() method and use it in the
checkValidityCommand() methods.

Extract method is a form of refactoring that enhances the readability
and extensibility of the code.

Refer to
https://nus-cs2103-ay2122s1.github.io/website/book/refactoring/what/
for an explanation of extract method.

BenjaminLHR committed:
Add exception handling for "find" commands

BenjaminLHR committed:
Fix bug in checkNumOfArguments() that returned erroneous output for
deadline and event commands.

BenjaminLHR committed:
Enhance readability by reorganizing sequence of methods and splitting
code into logical chunks using whitespace.

BenjaminLHR committed:
Merge branch 'branch-A-CodeQuality'

BenjaminLHR committed:
Merge branch 'branch-A-Assertions'

BenjaminLHR committed:
Modify message for miscellaneous DukeExceptions

BenjaminLHR committed:
Initialize DialogBox.fxml

BenjaminLHR committed:
Initialize MainWindow.fxml

BenjaminLHR committed:
Initialize MainWindow.java to serve as controller for MainWindow

BenjaminLHR committed:
Improve appearance of dialog boxes

BenjaminLHR committed:
Modify Main.java in accordance with the switch to fxml

BenjaminLHR committed:
Change JavaFX API from version 16 to 11

BenjaminLHR committed:
Merge remote-tracking branch 'origin/master'

BenjaminLHR committed:
Merge branch 'branch-Level-10'

BenjaminLHR committed:
Merge branch 'master' into branch-A-CodeQuality

BenjaminLHR committed:
Remove unnecessary imports

BenjaminLHR committed:
Add Javadoc comments.

BenjaminLHR committed:
Merge branch 'master' into branch-A-Assertions

BenjaminLHR committed:
Add assertion to verify that the exit status from Duke is either "0"
or "1".

BenjaminLHR committed:
Merge pull request #2 from BenjaminLHR/branch-A-CodeQuality

Improve code quality.

BenjaminLHR committed:
Merge branch 'master' into branch-A-Assertions

BenjaminLHR committed:
Merge pull request #3 from BenjaminLHR/branch-A-Assertions

Add assertion.

BenjaminLHR committed:
Bug fixes: Fix issue of incorrect index being displayed for FindCommand,
and issue of excess whitespace in task description causing
ArrayOutOfBoundsException

BenjaminLHR committed:
Fix bugs

BenjaminLHR committed:
Support shorter aliases for list, find, bye, todo, event commands

BenjaminLHR committed:
Merge branch 'branch-C-FriendlierSyntax'

BenjaminLHR committed:
Extract out portions of code that are similar to keep in line with DRY
principle.

BenjaminLHR committed:
Add user guide.

BenjaminLHR committed:
Set theme jekyll-theme-slate

BenjaminLHR committed:
Update _config.yml

BenjaminLHR committed:
Add screenshot of Duke

BenjaminLHR committed:
Merge remote-tracking branch 'origin/master'

BenjaminLHR committed:
Add markdown settings

BenjaminLHR committed:
Add exception checking for done and delete commands that throws
exception when the number of arguments is not one.

BenjaminLHR committed:
Remove unused variables and methods

BenjaminLHR committed:
Update "Setting Up in IntelliJ" portion of README

BenjaminLHR committed:
Fix bug of incorrect display when date of deadline contains "/by" or
date of event contains "/at"

BenjaminLHR committed:
Change regex used for segmentation in storage

BenjaminLHR committed:
Merge branch 'branch-changeStorageFormat'
